<div align="center">
<h1>Soal Latihan Dasar Kecerdasan Artifisial</h1>
<h3>Modul 2: Pengenalan Python</h3>
</div>

### Tujuan Praktikum
- Mahasiswa mengetahui instalasi dan *working environment* pada Python (menggunakan Terminal dan *Jupyter Notebook*)
- Mahasiswa memahamai cara pembuatan program dengan Python
- Mahasiswa mengetahui fitur-fitur utama yang tersedia pada Python

In [1]:
print("'Aarif Rahmaan Jalaluddin Faqiih") # Tuliskan nama lengkap anda
print("103112430182") # Tuliskan nim anda
# print("") # Tuliskan nomor meja
print("IF-12-07") # Tuliskan kelas anda
# print("") # Tuliskan kode asprak anda

'Aarif Rahmaan Jalaluddin Faqiih
103112430182
IF-12-07


### 1. Program Sederhana dengan Python (Bobot Nilai: `5`)

**a.** Buat program yang menampilkan kata "Selamat Belajar Python!" pada terminal dengan menggunakan `print()`

In [2]:
#jawab disini
print("Selamat Belajar Python!")

Selamat Belajar Python!


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |Selamat Belajar Python!|

**b.** buat program untuk menampilkan hasil dari 5 $(\sqrt{81})$. </br>
Tip: untuk menghitung akar pangkat 2 dari $x$, dapat menggunakan x**(1/2)

In [3]:
#jawab disini
print(5**(1/2))

2.23606797749979


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |45.0|

**c.** Program di bawah ini menghitung nilai dari persamaan kuadrat $8x^2 + 3x + 7$ dengan meminta *input* dari pengguan untuk variabel $x$. Minta *input* pada pengguna dengan deskripsi `"Masukkan nilai x: "`, kemudian ubah variabel `x` menjadi `integer`

In [4]:
# Minta input user dengan deskripsi yang sesuai
x = input("Masukkan nilai x: ")
# Ubah tipe data x menjadi integer
x = int(x)

# variabel ini langsung melakukan perhitungan didalamnya
hasil_persamaan = 8*x**2 + 3*x + 7

# Tampilkan hasilnya
print("Hasil dari persamaan 8x^2 + 3x + 7 dengan x =", x)
print("adalah", hasil_persamaan)

Masukkan nilai x: 4
Hasil dari persamaan 8x^2 + 3x + 7 dengan x = 4
adalah 147


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|Masukkan nilai x: `2` |45|
|Masukkan nilai x: `3` |88|
|Masukkan nilai x: `4`|147|

### 2.  Pengkondisian dalam Python (Bobot Nilai: `10`)

**a.** Buat program untuk mengecek apakah suatu bilangan adalah bilangan positif, negatif, atau nol. Jika bilangan tersebut positif maka tampilkan `"Bilangan Positif"` pada terminal, jika bilangan tersebut negatif maka tampilkan `"Bilangan Negatif"` pada terminal, dan jika bilangan tersebut nol maka tampilkan `"Bilangan Nol"` pada terminal.

In [5]:
# Minta input user dengan deskripsi yang sesuai
# karena input error, maka butuh tambahan builtins
nilai = __builtins__.input("Masukkan nilai: ")
# Ubah tipe data nilai menjadi integer
nilai = int(nilai)

if nilai > 0 :
    print(nilai, "adalah Bilangan Positif")
elif nilai < 0 :
    print(nilai, "adalah Bilangan Negatif")
else :
    print(nilai, "adalah Bilangan Nol")

Masukkan nilai: 6
6 adalah Bilangan Positif


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|Masukkan nilai: `5` |Bilangan Positif|
|Masukkan nilai: `-3` |Bilangan Negatif|
|Masukkan nilai: `0`|Bilangan Nol|

**b.** Buat program untuk meminta *input* usia dari pengguna. Program tersebut menerima usia dan menampilkan kategori usia pada terminal sesuai dengan kriteria berikut
- **Balita**: 0 - 5
- **Anak-anak**: 6 - 12
- **Remaja**: 13 - 17
- **Dewasa**: 18 - 59
- **Lansia**: 60 ke atas

In [6]:
# Minta input user dengan deskripsi yang sesuai
# karena input error, maka butuh tambahan builtins
usia = __builtins__.input("Masukkan usia: ")
# Ubah tipe data usia menjadi integer
usia = int(usia)

# Buat program di atas dengan menggunakan pengkondisian
if usia >= 0 and usia <= 5 :
    print("Usia", usia, "adalah Balita")
elif usia >= 6 and usia <= 12 :
    print("Usia", usia, "adalah Anak-anak")
elif usia >= 13 and usia <= 17 :
    print("Usia", usia, "adalah Remaja")
elif usia >= 18 and usia <= 59 :
    print("Usia", usia, "adalah Dewasa")
elif usia >= 60 :
    print("Usia", usia, "adalah Lansia")
else :
    print("Usia tidak valid")

Masukkan usia: 56
Usia 56 adalah Dewasa


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|Masukkan usia: `3` |Balita|
|Masukkan usia: `15` |Remaja|
|Masukkan usia: `65`|Lansia|

### 3. Prosedur dan Fungsi (Bobot Nilai: `10`)

**a.** Buatlah sebuah prosedur yang menerima parameter berupa nama dan jabatan. Jika jabatan yang diberikan adalah `"Manager"`, prosedur akan mencetak output `"Selamat pagi, Bapak/Ibu [nama]"`. Jika jabatan yang diberikan bukan "Manager", prosedur akan mencetak output `"Halo, [nama]"`. Sesuaikan output berdasarkan parameter yang diterima.

In [7]:
def sapa(nama, jabatan) :
  if jabatan == "manager" or jabatan == "manajer" :
    print("Selamat pagi, Bapak/Ibu", nama)
  else :
    print("Halo,", nama)

# Minta input user untuk nama dengan deskripsi yang sesuai
nama    = __builtins__.input("Masukkan nama: ")
# Minta input user untuk jabatan dengan deskripsi yang sesuai
jabatan_input = __builtins__.input("Masukkan jabatan: ")
# buat input jabatan menjadi lowercase agar tetap bisa diterima walaupun dalam case yang berantakan
jabatan = jabatan_input.lower()

# Panggil prosedur `sapa()` dengan parameter yang sesuai
sapa(nama, jabatan)

Masukkan nama: Faqiih
Masukkan jabatan: MaNaGeR
Selamat pagi, Bapak/Ibu Faqiih


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|Masukkan nama: `Budi`<br>Masukkan jabatan: `Staff` |Halo, Budi|
|Masukkan nama: `Dewi`<br>Masukkan jabatan: `Manager` |Selamat pagi, Bapak/Ibu Dewi|
|Masukkan nama: `Andi`<br>Masukkan jabatan: `Supervisor`|Halo, Andi|

**b.** Buatlah sebuah **fungsi** sederhana yang mengonversi suhu dari Celsius (C) menjadi Fahrenheit (F). Fungsi ini akan menerima input berupa suhu dalam Celsius, kemudian mengembalikan hasil konversi dalam Fahrenheit.

Rumus: $F = (C \times \frac{9}{5}) + 32$

In [8]:
def celcius_to_fahrenheit(celcius):
  fahrenheit = (celcius * 9/5) + 32
  return fahrenheit

celcius = float(__builtins__.input("Masukkan suhu (C): "))
fahrenheit = celcius_to_fahrenheit(celcius)
print("Dalam fahrenheit =", fahrenheit)

Masukkan suhu (C): 127
Dalam fahrenheit = 260.6


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|Masukkan suhu (C): `0` |32.0|
|Masukkan suhu (C): `100` |212.0|
|Masukkan suhu (C): `25`|77.0|

### 4. Perulangan (Bobot Nilai: `15`)

**a.** Buatlah program yang menerima *input* dari pengguna berupa sebuah bilangan bulat positif `n`. Program tersebut harus menampilkan semua bilangan genap dari 2 hingga `n` menggunakan perulangan `for` dan fungsi `range()`. </br>Tip: gunakan langkah kelipatan 2 dalam `range()` untuk menampilkan hanya angka genap.

In [9]:
# Buat kode untuk meminta input pengguna dan konversi input ke integer
n = int(__builtins__.input("Masukkan n: "))

# Perulangan untuk menampilkan angka genap sesuai input user
for i in range(2, n+1, 2) :
    print(i)

Masukkan n: 14
2
4
6
8
10
12
14


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|Masukkan n: `2` |2|
|Masukkan n: `7` |2<br>4<br>6|
|Masukkan n: `12`|2<br>4<br>6<br>8<br>10<br>12|

**b.** Buatlah sebuah fungsi yang menerima argumen `n` dan mengembalikan hasil perkalian (faktorial) dari bilangan bulat dari 1 hingga n. </br>Sebagai contoh, jika `n = 5`, fungsi tersebut harus menghitung perkalian **$1 \times 2 \times 3 \times 4 \times 5$** dan mengembalikan nilai 120. </br>Jika `n` kurang dari 1, kembalikan nilai 1. Gunakan perulangan `while` untuk menghitung hasil tersebut.

In [10]:
def faktorial(n): # Buat fungsi untuk menghitung faktorial
  if n < 1:       # Kondisi jika n kurang dari 1
    return 1
  else:
    hasil = 1
    i = 1
    while i <= n: # Buat kondisi untuk melakukan perulangan dari 1 hingga n
      hasil *= i  # Kalikan hasil dengan i
      i += 1      # Tambah i dengan 1 sebagai iterator

    return hasil  # Mengembalikan variabel yang menampung nilai hasil


# Buat kode untuk meminta input pengguna dan konversi input ke integer
user_input = int(input("Masukkan nilai n: "))

# Panggil fungsi untuk menghitung faktorial dan simpan hasilnya kedalam variabel
hasil_faktorial = faktorial(user_input)

# Tampilkan hasil faktorial yang sudah dihitung ke terminal
print("Hasil faktorial dari", user_input, "adalah", hasil_faktorial)

Masukkan nilai n: 9
Hasil faktorial dari 9 adalah 362880


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|Masukkan n: `5` |120|
|Masukkan n: `7` |5040|
|Masukkan n: `0`|1|

### 5. List dan Dictionary (Bobot Nilai: `20`)

**a.** Buat program yang mengganti status `"pending"` menjadi `"completed"` pada setiap elemen dalam list `tasks`. Gunakan metode `.replace()` untuk mengubah status, kemudian simpan status yang telah diubah ke dalam list baru bernama `updated_tasks` dan kemudian cetak list ke dalam terminal.

Tip :
```python
status_awal = "Task: pending"
print("Status awal:", status_awal)

status_baru = status_awal.replace("pending", "done")
print("Status baru:", status_baru)
```

output
```
Status awal: Task: pending
Status baru: Task: done
```

In [11]:
tasks = ["task1: pending", "task2: pending", "task3: pending", "task4: pending"]
updated_tasks = [] # Buat list kosong

for task in tasks : # Buat loop untuk iterasi elemen pada list tasks
    new_status = task.replace("pending", "completed") # Ganti status pending menjadi completed menggunakan .replace()
    updated_tasks.append(new_status) # Masukkan status yang sudah diganti ke dalam list updated_tasks menggunakan .append()

# Cetak list yang berisi status yang sudah diganti
print("Status baru:", updated_tasks)

Status baru: ['task1: completed', 'task2: completed', 'task3: completed', 'task4: completed']


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |['task1: completed', 'task2: completed', 'task3: completed', 'task4: completed']|

**b.** Buatlah fungsi `nomor_telepon` yang menerima sebuah dictionary. Dictionary ini berisi kode negara sebagai `key`, dan daftar nomor lokal sebagai `value`. Lengkapi bagian yang kosong dalam fungsi untuk menghasilkan daftar yang berisi nomor telepon lengkap. Nomor telepon ini dihasilkan dengan menggabungkan setiap nomor lokal dengan kode negara yang sesuai.

Tip:
```python
hobi_dict = {
    "Budi" : "Membaca",
    "Siti" : "Menulis",
    "Andi" : "Bermain Game"
}

for key, value in hobi_dict.items() :
    print("Key:", key, ", Value:", value)
```

Output
```
Key: Budi , Value: Membaca
Key: Siti , Value: Menulis
Key: Andi , Value: Bermain Game
```

In [12]:
def nomor_telepon(kode_negara) :
    telepon_lengkap = [] # List untuk menampung nomor telepon lengkap

    # Loop untuk setiap key dan value pada dict kode_negara
    for kode, daftar_nomor  in kode_negara.items() :
        for nomor in daftar_nomor:
            # Tambahakan nomor telepon lengkap ke list telepon_lengkap
            telepon_lengkap.append(kode+'-'+nomor)

    return telepon_lengkap

data_telepon = {
    "+62": ["81234567890", "82198765432", "85312345678"],
    "+1" : ["2025551234", "3105559876"],
    "+44": ["7911123456", "7700900123"]
}

# Panggil fungsi nomor_telepon, kemudian masukkan ke dalam variabel
hasil = nomor_telepon(data_telepon)

# Cetak list hasil nomor telepon lengkap yang didapatkan dari nomor_telepon
print(hasil)

['+62-81234567890', '+62-82198765432', '+62-85312345678', '+1-2025551234', '+1-3105559876', '+44-7911123456', '+44-7700900123']


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |['+62-81234567890', '+62-82198765432', '+62-85312345678', '+1-2025551234', '+1-3105559876', '+44-7911123456', '+44-7700900123']|

### 6. Soal Cerita (Bobot Nilai: `40`)

Seorang pemilik perpustakaan ingin membuat program sederhana untuk mencatat **peminjaman buku harian** oleh anggota. Setiap anggota akan dicatat jumlah buku yang dipinjam selama **tiga waktu** dalam sehari (pagi, siang, sore). Pemilik perpustakaan perlu menghitung **total buku yang dipinjam** untuk setiap anggota dan menentukan apakah anggota tersebut mendapat **diskon** atau **tidak**. Kriteria diskon adalah jika total buku yang dipinjam sama dengan atau lebih dari **10 buku**.

Buatlah program yang melakukan hal berikut :
1. **Input/Output**
    - Meminta input dari pengguna berupa `nama anggota` dan `jumlah buku yang dipinjam` untuk tiga waktu (pagi, siang, sore)
    - Menampilkan **total peminjaman** dan **status diskon** setiap anggota

2. **Penentuan diskon**
    - Menentukan apakah anggota **dapat diskon** atau **tidak dapat diskon** berdasarkan total buku yang dipinjam.

3. **Pengulangan program**
    - Menggunakan perulangan untuk memasukkan data beberapa anggota hingga pengguna memasukkan `"selesai"` sebagai nama anggota

4. **Perhitungan total**
    - Membuat fungsi untuk **menghitung total** yang menerima list peminjaman sebagai parameter dan mengembalikan total buku dari list tersebut

5. **Penyimpanan data**
    - Gunakan *dictionary* dengan `keys` nama anggota, dan `value` list peminjaman dari anggota tersebut.

In [13]:
# Fungsi untuk menghitung total peminjaman
def hitung_total(daftar_peminjaman):
  return sum(daftar_peminjaman)

# Dictionary untuk menyimpan data anggota
data_anggota = {}

# Input data
while True:
  nama = input("Masukkan nama anggota (atau 'selesai' untuk berhenti): ")

  if nama.lower() == "selesai":
    break

    # Input jumlah buku untuk tiga waktu
  pagi  = int(input("Masukkan jumlah buku pagi: "))
  siang = int(input("Masukkan jumlah buku siang: "))
  sore  = int(input("Masukkan jumlah buku sore: "))

  # Simpan dalam dictionary sebagai list
  data_anggota[nama] = [pagi, siang, sore]

print("\n=== Rekap Peminjaman Anggota ===")

if not data_anggota:
  print("(tidak ada data anggota)")
else:
  for nama, peminjaman in data_anggota.items():
    total = hitung_total(peminjaman)

    if total >= 10:
      status = "Dapat Diskon"
    else:
      status = "Tidak Dapat Diskon"

    print(f"Nama: {nama} | Peminjaman: {peminjaman} | Total: {total} buku | Status: {status}")

Masukkan nama anggota (atau 'selesai' untuk berhenti): Heri
Masukkan jumlah buku pagi: 3
Masukkan jumlah buku siang: 4
Masukkan jumlah buku sore: 3
Masukkan nama anggota (atau 'selesai' untuk berhenti): Hartono
Masukkan jumlah buku pagi: 5
Masukkan jumlah buku siang: 6
Masukkan jumlah buku sore: 4
Masukkan nama anggota (atau 'selesai' untuk berhenti): Sandra
Masukkan jumlah buku pagi: 1
Masukkan jumlah buku siang: 2
Masukkan jumlah buku sore: 2
Masukkan nama anggota (atau 'selesai' untuk berhenti): William
Masukkan jumlah buku pagi: 4
Masukkan jumlah buku siang: 3
Masukkan jumlah buku sore: 4
Masukkan nama anggota (atau 'selesai' untuk berhenti): Risa
Masukkan jumlah buku pagi: 2
Masukkan jumlah buku siang: 1
Masukkan jumlah buku sore: 3
Masukkan nama anggota (atau 'selesai' untuk berhenti): selesai

=== Rekap Peminjaman Anggota ===
Nama: Heri | Peminjaman: [3, 4, 3] | Total: 10 buku | Status: Dapat Diskon
Nama: Hartono | Peminjaman: [5, 6, 4] | Total: 15 buku | Status: Dapat Diskon
Na

**Bagian yang diberikan tanda `ini` adalah masukkan user**

*Test case* program

**|----------------------------------------------------------------------------------|**

**Test Case 1: Dua anggota dengan status berbeda**

```
Masukkan nama anggota (atau 'selesai' untuk berhenti): `Ahmad`
Masukkan jumlah buku pagi: `4`
Masukkan jumlah buku siang: `3`
Masukkan jumlah buku sore: `5`

Masukkan nama anggota (atau 'selesai' untuk berhenti): `Bella`
Masukkan jumlah buku pagi: `2`
Masukkan jumlah buku siang: `1`
Masukkan jumlah buku sore: `3`

Masukkan nama anggota (atau 'selesai' untuk berhenti): `selesai`

=== Rekap Peminjaman Anggota ===
Nama: Ahmad | Peminjaman: [4, 3, 5] | Total: 12 buku | Status: Dapat Diskon
Nama: Bella | Peminjaman: [2, 1, 3] | Total: 6 buku | Status: Tidak Dapat Diskon
```

**|----------------------------------------------------------------------------------|**

**Test Case 2: Anggota dengan total peminjaman tepat di batas**

```
Masukkan nama anggota (atau 'selesai' untuk berhenti): `Citra`
Masukkan jumlah buku pagi: `3`
Masukkan jumlah buku siang: `4`
Masukkan jumlah buku sore: `3`

Masukkan nama anggota (atau 'selesai' untuk berhenti): `selesai`

=== Rekap Peminjaman Anggota ===
Nama: Citra | Peminjaman: [3, 4, 3] | Total: 10 buku | Status: Dapat Diskon
```

**|----------------------------------------------------------------------------------|**

**Test Case 3: Langsung selesai tanpa input anggota**

```
Masukkan nama anggota (atau 'selesai' untuk berhenti): `selesai`

=== Rekap Peminjaman Anggota ===
(tidak ada data anggota)
```

**|----------------------------------------------------------------------------------|**

**Test Case 4: Tiga anggota dengan berbagai kondisi**

```
Masukkan nama anggota (atau 'selesai' untuk berhenti): `Dimas`
Masukkan jumlah buku pagi: `5`
Masukkan jumlah buku siang: `6`
Masukkan jumlah buku sore: `4`

Masukkan nama anggota (atau 'selesai' untuk berhenti): `Eka`
Masukkan jumlah buku pagi: `1`
Masukkan jumlah buku siang: `2`
Masukkan jumlah buku sore: `2`

Masukkan nama anggota (atau 'selesai' untuk berhenti): `Farhan`
Masukkan jumlah buku pagi: `4`
Masukkan jumlah buku siang: `3`
Masukkan jumlah buku sore: `4`

Masukkan nama anggota (atau 'selesai' untuk berhenti): `selesai`

=== Rekap Peminjaman Anggota ===
Nama: Dimas | Peminjaman: [5, 6, 4] | Total: 15 buku | Status: Dapat Diskon
Nama: Eka | Peminjaman: [1, 2, 2] | Total: 5 buku | Status: Tidak Dapat Diskon
Nama: Farhan | Peminjaman: [4, 3, 4] | Total: 11 buku | Status: Dapat Diskon
```